platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/exec/BuildTreeLifecycleBuildActionExecutor.java

```diff
@@ -78,16 +89,28 @@ public BuildActionRunner.Result execute(BuildAction action, BuildSessionContext
         return result;
     }
 
-    private static BuildActionModelRequirements buildActionModelRequirementsFor(BuildAction action) {
+    private BuildActionModelRequirements buildActionModelRequirementsFor(BuildAction action) {
         if (action instanceof BuildModelAction && action.isCreateModel()) {
             BuildModelAction buildModelAction = (BuildModelAction) action;
-            return new QueryModelRequirements(action.getStartParameter(), action.isRunTasks(), buildModelAction.getModelName());
+            Object payload = buildModelAction.getModelName();
+            return new QueryModelRequirements(action.getStartParameter(), action.isRunTasks(), payloadHashProvider(payload));
         } else if (action instanceof ClientProvidedBuildAction) {
-            return new RunActionRequirements(action.getStartParameter(), action.isRunTasks());
+            SerializedPayload payload = ((ClientProvidedBuildAction) action).getAction();
+            return new RunActionRequirements(action.getStartParameter(), action.isRunTasks(), payloadHashProvider(payload));
         } else if (action instanceof ClientProvidedPhasedAction) {
-            return new RunPhasedActionRequirements(action.getStartParameter(), action.isRunTasks());
+            SerializedPayload payload = ((ClientProvidedPhasedAction) action).getPhasedAction();
+            return new RunPhasedActionRequirements(action.getStartParameter(), action.isRunTasks(), payloadHashProvider(payload));
         } else {
             return new RunTasksRequirements(action.getStartParameter());
         }
     }
+
+    private Supplier<HashCode> payloadHashProvider(Object payload) {
+        ValueSnapshotter valueSnapshotter = this.valueSnapshotter;
+        return () -> {
+            Hasher hasher = Hashing.newHasher();
```

Comment:
🤔 This is the exact same code implemented by `ToolingModelParameterCarrier` in that other PR. Do you plan to dedupe them? Perhaps this could even be a default method in `ValueSnapshotter`?

Reply:
It is a small part of the code, that I don't think deserves to be deduplicated via a shared function.

I made it more concise by introducing a helper on `Hashing` that serves a common need. I wouldn't want to change the `ValueSnapshotter` interface for this.